Add the ability for Pants to provide Python via a union (with a pyenv impl)

thejcannon opened the pull request:
Fixes #10447

Adding the ability for an optional union (`PythonProvider`) implementer to return a `PythonExecutable` (newly outfitted with `append_only_caches`) so that Python for user code can be supplied completely by rule-code. Additionally added a `pyenv` union implementation.

The dirty details are:
- Python is essentially NOT relocatable (hats off to IndyGreg). Python itself, as well as ecosystem tools, are baked with assumptions about absolute paths.
- Python installation is not atomic. That means we do our work behind a lock. We run the bash script on every session, but it should be fast as it only does the work if the target python doesnt exist.

thejcannon commented on src/python/pants/init/plugin_resolver.py:
```diff
@@ -80,7 +80,6 @@ async def resolve_plugins(
             internal_only=True,
             python=python,
             requirements=requirements,
-            interpreter_constraints=request.interpreter_constraints,
```
This was `None` and thats a no-no. However there's no type hint for this param so `mypy` isn't complaining.

thejcannon commented:
We still require some kind of `python` for executing non-user code. Namely:
- `pex` to ask about ICs
- gunzipping, and a few other places

I can tackle that in another PR.

kaos commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,182 @@
+    pass
+
+
+PYENV_APPEND_ONLY_CACHES = FrozenDict({"pyenv": ".pyenv"})
```
would like to see the use of `.pyenv` codified into a CONST or similar to a) make the dependency clear where this is used and b) to avoid inadvertent changes to a subset of them.

kaos commented on src/python/pants/backend/python/providers/pyenv/rules_integration_test.py:
```diff
@@ -0,0 +1,122 @@
+def run_run_request(
+    rule_runner: RuleRunner,
+    target: Target,
+) -> Tuple[int, str, str]:
```
using a `namedtuple` would make this less convoluted on the receiving end.

thejcannon replied:
I stole this from the run tests 🤷‍♂️

kaos commented on src/python/pants/backend/python/providers/pyenv/rules_integration_test.py:
```diff
@@ -0,0 +1,122 @@
+    """A VenvPex refers to the location of the venv so it doesn't have to re-construct if it exists.
+
+    Part of this location is a hash of the interpreter. Without careful consideration it can be easy
+    for this hash to drift from build-time to run-time. This invalidates the assumption that the
```
invalidates? looks like it validates..?

kaos commented on src/python/pants/backend/python/providers/pyenv/rules_integration_test.py:
```diff
@@ -0,0 +1,122 @@
+    named_caches_dir = (
+        rule_runner.options_bootstrapper.bootstrap_options.for_global_scope().named_caches_dir
+    )
+    assert stdout.startswith(f"{named_caches_dir}/pyenv")
```
I was hoping/expecting to see an assert that the actual Python used matches the constraint given.

thejcannon replied:
That's a good idea

kaos reviewed (commented):
This is huge. Amazing stuff!

kaos commented:
> That means we do our work in a tempdir and symlink the result, on every session.

how much needs to be redone though? would be nice if we could avoid reinstalling Python on every restart of pantsd (that's what session means, right? or is it every pants run? cause that sounds like it would be a little bit of a buzzkill)

thejcannon commented:
Oh sorry bad phrasing. We run the process on every run, but bail if the python already exists. So we should only install it once. I'll fix the phrasing.

kaos reviewed (approved):
🚀

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+    #   fastest-yet-still-correct solution is to always run this process and make it bail
+    #   early-and-quickly if the requisite Python already exists.
+    #   2. Pyenv compiles Python using whatever compiler the system is configured to use. Python
+    #   then stores this information so that it cn use the same compiler when compiling extension
```
```suggestion
    #   then stores this information so that it can use the same compiler when compiling extension
```

But I wouldn't trigger a new CI run to fix this. Unless you make more changes in this PR, just tack it on to some unrelated future PR.

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+    )
+
+    return PythonExecutable(
+        path=result.stdout.decode("utf-8").splitlines()[-1].strip(),
```
Just `decode()` will use UTF-8, and that seems to be how we do this idiomatically elsewhere in the codebase.

In fact, you could use `encode()` and `decode()` above and drop the `"ascii"`, since UTF-8 is a superset of ascii and you don't expect any non ascii characters...

benjyw replied:
But again, not worth triggering a new CI run for.

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+        return f"./pyenv-{self.version}/bin/pyenv"
+
+
+class PythonToolchainRequest(PythonProvider):
```
It's not really a "toolchain", right? Just an interpreter? It doesn't even have setuptools or wheel installed, necessarily? Maybe wtih should be `PyenvPythonProvider`, for clarity?

thejcannon replied:
Ahh sorry, carryover from the initial parts of this change, which used the word "Toolchain" (borrowed from Bazel's)

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+            input_digest=pyenv.digest,
+            description=f"Choose specific version for Python {python_to_use}",
+            env={"PATH": env_vars.get("PATH", "")},
+            # Caching the result is OK, since if the user really needs a different patch,
```
I concur.  And this will only matter when new Python versions are released, which is... not often.

thejcannon commented:
Another open question: Should we start using this in this repo? There's two places it could be useful:
- For running the repo itself `./pants run ...`
- For our tests that are multi-interpreter. Right now we skip, but with enabling the `pyenv` plugin in the test, we wouldn't have to (although there'd be timeout issues :grimacing: )

benjyw commented:
Awesome! Users will love it. Reviewing now.

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+                        #!/usr/bin/env bash
+                        set -e
+                        DEST=$(realpath {PYENV_NAMED_CACHE}/{specific_python})
+                        if [ ! -d "$DEST" ]; then
```
I guess if a user only partially wipes this dir, you're asking for trouble, and the remedy is to fully wipe it.

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+                        set -e
+                        DEST=$(realpath {PYENV_NAMED_CACHE}/{specific_python})
+                        if [ ! -d "$DEST" ]; then
+                            export PYENV_ROOT=$(realpath $(mktemp -d -u -p {PYENV_NAMED_CACHE} {specific_python}.XXXXXX))
```
OK so we create an ad-hoc PYENV_ROOT for each install (which is why specific_python is mixed into its name)? That is not idiomatic for pyenv, which normally keeps all its versions in the same root, but it makes sense here I think.

thejcannon replied:
Yup. And that is because:

> Python installation is not atomic.

So if we did just set `PYENV_ROOT` to the named cache dir, the resulting directory/binary would exist before its "ready".
Alternatively we could `touch PYENV_ROOT/versions/{version}/DONE` but that means `pyenv` installations in parallel have to be OK. They might be but IDK if they are :unamused: 

I'm very eager to make this cleaner, so feel free to brainstorm as much as you want with me :smile:

benjyw replied:
Eh, I think it's fine... :)

benjyw reviewed (approved):
Potentially massive!!

benjyw commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+            # they should list a more precise IC.
+        ),
+    )
+    specific_python = which_python_result.stdout.decode("ascii").strip()
```
Can remove the "ascii" here too

thejcannon commented:
Oh one alternative to the current bash script is a Python script which uses locks and subprocesses. If we fix the issue highlighted above regarding non-user-code usage of Python (ideally with `pyoxy` or Python Build Standalone) a Python script to install Python is possible.

thejcannon commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+    default_version = "2.3.13"
+    default_url_template = "https://github.com/pyenv/pyenv/archive/refs/tags/v{version}.tar.gz"
+
+    python_configure_opts = StrListOption(
+        help="Flags to use when configuring CPython.",
+        advanced=True,
+    )
```
This doesn't invalidate the existing Python. That's a choice I made. Should it though?

benjyw replied:
Probably? Also, is this actually limited to cpython? Seems like I could specify other pythons in my ICs?

thejcannon replied:
This specific flag is limited to CPython: https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#special-environment-variables

If we wanted it to invalidate, we'd need to make the path have a hash of this data.

thejcannon replied:
I removed this option. `--enable-optimizations` is SUPER slow.

thejcannon commented:
The failing test is kinda invalid... it relies on changing the `interpreter_constraints` to `PluginResolver`, which in "production" is always `None`. But that's for another day...

kaos commented:
> (I'm surprised there's no issue for this)

turns out there is: #10447

thejcannon commented:
Ah, another note too. I should mark this experimental so we don't feel pressured to get it completely right the first go-around.

thejcannon commented:
> Oh one alternative to the current bash script is a Python script which uses locks and subprocesses. If we fix the issue highlighted above regarding non-user-code usage of Python (ideally with `pyoxy` or Python Build Standalone) a Python script to install Python is possible.

I'm gonna do this... I think...

benjyw commented:
Why? A bash script seems fine to me.

thejcannon commented:
Well I thought Python would make the locking easier. But I _think_ I got it working in bash.

Either way, my comment wasn't 100% accurate. We will always have some kind of python binary laying around and isn't dependent on some future PR. Although it'd be sweet if that binary was also provided by Pants (spoiler: it will be)

thejcannon commented:
I don't think it's worth marking experimental FWIW.

Last little question is if we should set the default flags to be [compiling with maximum performance](https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-for-maximum-performance).

thejcannon commented:
OK we definitely should not, lol. I'm at 600 seconds and counting.

thejcannon commented:
> Well I thought Python would make the locking easier. But I think I got it working in bash.

Oh I should just use `flock` I think...

jsirois commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,183 @@
+                                touch "$DONE"
+                            fi
+                        fi
+                        echo $(realpath "$DEST"/bin/python)
```
I said I'd stay out, but no one seems to have noticed this.

So, you have a problem:
```
$ ssh gha@pantsbuild-gha-mac-M1 realpath --version
zsh:1: command not found: realpath
$ ssh gha@pantsbuild-gha-mac-M1 flock --version
zsh:1: command not found: flock
$ ssh gha@pantsbuild-gha-mac-x86-64 realpath --version
zsh:1: command not found: realpath
$ ssh gha@pantsbuild-gha-mac-x86-64 flock --version
zsh:1: command not found: flock
```

This is not even to mention if the machine doesn't have the pyenv Python build requirements. The failures in these cases are likely pretty cryptic. It would be good to make sure that, from the perspective of someone running `./pants ...` who doesn't have one or more of these things, that the error is semi-grokkable.

thejcannon replied:
Hmm correct me if I'm wrong, but wouldn't the `#!/usr/bin/env bash` at the top mean your OS would run the script through bash? Those lines seem like it's `zsh`.

Agreed on the build deps, but not sure the best way to handle. (I wonder if the pyenv/Python error is already grokkable?)

jsirois replied:
They are not a bash builtins either - they're binaries on your system or not independent of the shell. Re-running with bash -c '...' gives the same answer - Macs don't have these binaries by default.

thejcannon replied:
Okie dokie, looks like I'll be swtiching to a Python-based script then. Thanks for the callout! Bummer the tests aren't being run on Mac (it seems :thinking: )

jsirois replied:
I think `@pytest.mark.platform_specific_behavior` is required for that.

jsirois replied:
But don't let that fool you either. It means ~nothing. GHA Macs have alot of non-default software installed, for example. The only real solution when using shell to implement Pants goals is to research well.

benjyw replied:
That seems odd though - you need python to install python? Then I can just ensure the right python up front and not need any of this.

jsirois replied:
You should be able to use `sys.executable` to run a Python script that just uses `subprocess` and `fcntl` just fine. It's an abstraction leak to use the Pants host Python like this; so it should be called out in a comment.

jsirois replied:
But you could also just implement realpath with readlink I think and go back to the polling solution.

thejcannon replied:
Thanks for the tip, but I've taken your Slack comment(s) to heart and just going to assume `sys.executable` is off limits.

> That seems odd though - you need python to install python? Then I can just ensure the right python up front and not need any of this.

We already have this restriction today. There must be _some_ kind of Python that exists so we can run non-user code. The two main instances being rule-code that needs to `gunzip` and `pex` itself (to select ICs). Look for the `find_python` rule.

Whether that code continues as is, becomes `sys.executable`, or something else is up to a future PR :smile:

benjyw replied:
`sys.executable` should be off-limits for user code, but this is not that.

thejcannon replied:
Either way that's an implementation detail of `find_python`, and we could discuss that _not_ in this PR, as this PR will just use `PythonBinary` :wink:

benjyw replied:
We can discuss in a separate issue, but note that PythonBinary uses the `python-bootstrap` subsystem to find python, and that is tangled up with interpreter discovery for user code. Specifically [here](https://github.com/benjyw/pants/blob/a8d9179426c2e46a1f368e04b869a66df486026f/src/python/pants/backend/python/util_rules/pex_environment.py#L174). 

Probably `PythonBinary` should just be hardwired to `sys.executable` and bypass all the rest of it.

thejcannon replied:
Yeah that's my point 😀

thejcannon replied:
Actually, no we can't. `sys.executable` is for the environment host :smile:

thejcannon commented:
Okie dokie folks, hot changes coming in:
- Switched the locking and installation to Python to be cross-platform(-ish). The sentinel file-checking is still in bash because it's the hot-path and bash go vroom.
- Added a bunch of plumbing to make it so users can out-of-band install versions the same way we do in-band
  - We now have a new (internal) target type which is runnable representing running `pyenv install`.
  - And a synthetic target is constructed at the root named `pyenv-install`
  - This should be used for people to do more complicated things, like installing an optimized version of Python
- Added more help docs

kaos commented on src/python/pants/backend/python/providers/pyenv/target_types.py:
```diff
@@ -0,0 +1,17 @@
+class PyenvInstall(Target):
+    alias = "_pyenv_install"
+    help = "<internal target>"
+    core_fields = (PyenvInstallSentinelField,)
```
this is a bit boiler-platey. not suggesting to do anything about it here and now, but if we find this pattern useful, it could make sense to streamline it a bit :)

kaos commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,336 @@
+
+            sudo rm -rf <named_caches_dir>/pyenv/versions/<specific_version>
+            # env vars from https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-for-maximum-performance
+            PYTHON_CONFIGURE_OPTS='--enable-optimizations --with-lto' {bin_name()} run :pyenv-install -- 3.10
```
This is cool. We effectively get "free" sub-commands to our goals with synthetic targets!

kaos replied:
your last commit made this comment stale.

thejcannon replied:
fixed

kaos reviewed (approved):
Awesome 🚀

stuhood commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,338 @@
+    #   runs (technically during a run, but we can't do anything about that) the
+    #   fastest-yet-still-correct solution is to always run this process and make it bail
+    #   early-and-quickly if the requisite Python already exists.
+    #   2. Pyenv compiles Python using whatever compiler the system is configured to use. Python
+    #   then stores this information so that it can use the same compiler when compiling extension
+    #   modules. Therefore caching the compiled Python is somewhat unsafe (especially for a remote
+    #   cache).
```
Could reference https://github.com/pantsbuild/pants/issues/10769 here probably.

stuhood reviewed (approved):
Nice! Thanks.

As to Docker: this should work fine, since named caches have been used properly AFAICT.

stuhood commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,338 @@
+    name = "pyenv"
+    help = softwrap(
+        f"""
+        A subsystem for pyenv-provided-Python (https://github.com/pyenv/pyenv).
+
+        Enabling this subsystem will switch Pants from trying to find an appropriate Python on your
+        system to using pyenv to install the correct Python(s).
```
...Which also highlights that some doc changes might be good for this, so that those two strategies reference one another.

thejcannon replied:
Ahh yeah, docs always elude me. I'll see what I can do

thejcannon replied:
I'm gonna punt on docs changes in this PR and open another one for that.

thejcannon commented on src/python/pants/backend/python/providers/pyenv/rules.py:
```diff
@@ -0,0 +1,338 @@
+        process. Doing so would increase the time it takes to install a single Python by about an
+        order of magnitude (E.g. ~2.5 minutes to ~26 minutes).
+
+        If you wish to customize the pyenv installation of a Python, a synthetic target is exposed
+        at the root of your repo which is runnable. This target can be run with the relevant
+        environment variables set to enable optimizations. You will need to wipe the specific
+        version directory if Python was already installed. Example:
+
+            sudo rm -rf <named_caches_dir>/pyenv/versions/<specific_version>
+            # env vars from https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-for-maximum-performance
+            PYTHON_CONFIGURE_OPTS='--enable-optimizations --with-lto' {bin_name()} run :pants-pyenv-install -- 3.10
```
If we expose this through options it's the same end result. Customized python builds...

And this way is the only "middle ground" I could come up with to make having some (uncacheable) build step take 30 minutes. Make the user "own" running that.

thejcannon replied:
Well maybe I'm just viewing this from my perspective, but I know my engineers wouldn't wait for a 30 minute process. And they'd bug me constantly about it. And kill it. Meaning they'd have to restart.

So I can't put it in pants.toml. 

And putting it there feels weird because if you change the option we don't invalidate existing pythons either. (Same for .pants.rc)

thejcannon replied:
All in all, the end goal is providing Python. That's the 80% case.

For the 20%, let's give them a back door. But it doesn't have to be of the same elk as the front door 😉

stuhood replied:
> So I can't put it in pants.toml.

This is true for optimization flags, but not true for the other flags you might want to pass to a `pyenv` install: particularly `CONFIGURE_OPTS`.

> And putting it there feels weird because if you change the option we don't invalidate existing pythons either. (Same for .pants.rc)

We could? Fingerprinting the options could be done by embedding them in the installation script, and taking its digest, for example.

thejcannon replied:
Sure anything is possible. Getting back to it though, the build is already non-hermetic and we still need a way to allow users to install Python out-of-band or suffer a 30-minute build step.

thejcannon replied:
OK I think I've decided to have this functionality be in its own sub-plugin of the `pyenv` provider plugin. So it's "off" by default.

thejcannon commented on src/python/pants/core/goals/run.py:
```diff
@@ -132,8 +132,8 @@ def __init__(
         object.__setattr__(self, "extra_env", FrozenDict(extra_env or {}))
-        object.__setattr__(self, "immutable_input_digests", immutable_input_digests)
-        object.__setattr__(self, "append_only_caches", append_only_caches)
+        object.__setattr__(self, "immutable_input_digests", FrozenDict(immutable_input_digests))
+        object.__setattr__(self, "append_only_caches", FrozenDict(append_only_caches))
```
This was likely cruft from removing `frozen_after_init`

kaos replied:
to avoid superflous object creation in case the inputs already are frozen dicts could do
```suggestion
        object.__setattr__(self, "immutable_input_digests", FrozenDict.freeze(immutable_input_digests))
        object.__setattr__(self, "append_only_caches", FrozenDict.freeze(append_only_caches))
```

thejcannon replied:
This wouldn't mirror the `FrozenDict` call above. I'd rather be consistent locally and we address `.freeze` across the board.

thejcannon commented:
Marked as experimental so we can merge with a few conversations ongoing.
